### PR TITLE
Make email input on login/registration less sensitive to spaces

### PR DESCRIPTION
**Pull request by TheReshma**
## Trimming whitespace at the beginning & end of the email as the user types.

This solves the Issue [#20239](https://github.com/vector-im/element-web/issues/20239) in Element-Web.

### Before : 

Leaving whitespace before or after the email raises an "Invalid Email" alert because the email field is space sensitive.
image

### After : 

Whitespace at the beginning & end of the email is trimmed as the user types. Thus solving the issue.
image

Signed-off-by: Reshma <reshmashaik3011@gmail.com>

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Make email input on login/registration less sensitive to spaces ([\#8220](https://github.com/matrix-org/matrix-react-sdk/pull/8220)). Contributed by @TheReshma.

### CODE REVIEW

File: src/components/structures/auth/ForgotPassword.tsx [hunk 1]
```diff
@@ -287,7 +287,7 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
                         name="reset_email" // define a name so browser's password autofill gets less confused
                         labelRequired={_td('The email address linked to your account must be entered.')}
                         labelInvalid={_td("The email address doesn't appear to be valid.")}
-                        value={this.state.email}
+                        value={(this.state.email).trim()}
                         fieldRef={field => this[ForgotPasswordField.Email] = field}
                         autoFocus={true}
                         onChange={this.onInputChanged.bind(this, "email")}
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:293]: why not handle the trim in the `onInputChanged` function, so we can have a comment for it?

Comment (TheReshma) [reply src/components/structures/auth/ForgotPassword.tsx]: Thank you. Made the changes by using `trim()` in the `onInputChanged` function.

File: src/components/structures/auth/ForgotPassword.tsx [hunk 2]
```diff
@@ -225,7 +225,7 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
 
     private onInputChanged = (stateKey: string, ev: React.FormEvent<HTMLInputElement>) => {
         this.setState({
-            [stateKey]: ev.currentTarget.value,
+            [stateKey]: (ev.currentTarget.value).trim(),
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:228]: we shouldn't be applying this to the password fields though... we'll need an `if` statement.

**Review by turt2live — CHANGES_REQUESTED** [review]

File: src/components/structures/auth/ForgotPassword.tsx (hunk 2 above)
Comment (TheReshma) [reply src/components/structures/auth/ForgotPassword.tsx]: ###  Added an If Statement to check if it is the `email` field or not.

```
 private onInputChanged = (stateKey: string, ev: React.FormEvent<HTMLInputElement>) => {
        if( stateKey == "email"){
            this.setState({
                [stateKey]: (ev.currentTarget.value).trim(),
            }as any);
        }else{
            this.setState({
                [stateKey]: ev.currentTarget.value,
            }as any);
        } 
    };
```

Let me know if there are further changes, Thanks !

**Comment by andybalaam** [comment]
@TheReshma some of the builds are failing. When I click "Details" above I see these errors:

```
 /workdir/src/components/structures/auth/ForgotPassword.tsx
  227:9   error  Expected space(s) after "if"               keyword-spacing
  227:12  error  There should be no space after this paren  space-in-parens
  227:33  error  Missing space before opening brace         space-before-blocks
  231:10  error  Expected space(s) before "else"            keyword-spacing
  231:10  error  Expected space(s) after "else"             keyword-spacing
  235:10  error  Trailing spaces not allowed                no-trailing-spaces
```

**Comment by TheReshma** [comment]
I made the changes & ran `yarn lint` . It shows many errors relating to exports and types but none related to the section I am working on.

File: src/components/structures/auth/ForgotPassword.tsx [hunk 3]
```diff
@@ -222,9 +222,15 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
     }
 
     private onInputChanged = (stateKey: string, ev: React.FormEvent<HTMLInputElement>) => {
-        this.setState({
-            [stateKey]: ev.currentTarget.value,
-        } as any);
+        if (stateKey == "email") {
+            this.setState({
+                [stateKey]: (ev.currentTarget.value).trim(),
+            }as any);
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:228]:
```suggestion
            } as any);
```

File: src/components/structures/auth/ForgotPassword.tsx [hunk 4]
```diff
@@ -222,9 +222,15 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
...
+        } else {
+            this.setState({
+                [stateKey]: ev.currentTarget.value,
+            }as any);
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:232]:
```suggestion
            } as any);
```

File: src/components/structures/auth/ForgotPassword.tsx [hunk 5]
```diff
@@ -222,9 +222,15 @@ export default class ForgotPassword extends React.Component<IProps, IState> {
     }
 
     private onInputChanged = (stateKey: string, ev: React.FormEvent<HTMLInputElement>) => {
-        this.setState({
-            [stateKey]: ev.currentTarget.value,
-        } as any);
+        if (stateKey == "email") {
```
Comment (turt2live) [inline src/components/structures/auth/ForgotPassword.tsx:225]:
```suggestion
        if (stateKey === "email") {
```

**Review by turt2live — CHANGES_REQUESTED** [review]
otherwise looks good - thanks! Another approach would be to do this:

```typescript
let value = ev.currentTarget.value;
if (stateKey === "email") value = value.trim();
this.setState({
    [stateKey]: value,
} as any);
```

**Comment by turt2live** [comment]
It also looks like you've made this PR off your `develop` branch - I'd highly suggest closing this and opening a new PR from a dedicated branch to fix the CI/tests.

**Comment by TheReshma** [comment]
> It also looks like you've made this PR off your `develop` branch - I'd highly suggest closing this and opening a new PR from a dedicated branch to fix the CI/tests.

Sure

**Pull request closed without merging**